strawberryfields/decompositions.py

```diff
@@ -400,44 +414,139 @@ def mach_zehnder(m, n, internal_phase, external_phase, nmax):
-    interferometers consisting of two phase-shifters and two symmetric (50:50)
-    beamsplitters.
+    if U[m, n] == 0:
+        phi_i = np.pi
+        phi_e = 0
+    elif U[m, n + 1] == 0:
+        phi_i = 0
+        phi_e = 0
+    else:
+        r = -U[m, n + 1] / U[m, n]
+        phi_i = 2 * np.arctan(np.abs(r))
+        phi_e = -np.angle(r)
 
-    The two beamsplitters in this decomposition of T are modeled by :class:`~.ops.BSgate`
-    with arguments :math:`(\pi/4, \pi/2)`, and the two phase-shifters (see :class:`~.ops.Rgate`)
-    act on the input mode with the lower index of the two. The phase imposed
-    by the first phaseshifter (before the first beamsplitter) is named
-    ``external_phase``, while we call the phase shift between the beamsplitters
-    ``internal_phase``.
+    return [n, n + 1, phi_i, phi_e, nmax]
 
-    The algorithm applied in this function makes use of the following identity:
 
-    .. code-block:: python
+def nullMZ(n, m, U):
+    r"""Nullifies element n,m of U using mach_zehnder.
 
-        Rgate(alpha) | 1
-        Rgate(beta) | 2
-        Rgate(phi) | 1
-        BSgate(theta, 0) | 1, 2
+    Args:
+        n (int): row index of element to be nullified
+        m (int): column index of element to be nullified
+        U (array): matrix whose m,n element is to be nullified
 
-        equals
+    Returns:
+        list: list containing ``[m, n, internal_phase, external_phase, nmax]`` of the
+            mach_zehnder unitaries needed
+    """
+    (nmax, mmax) = U.shape
 
-        Rgate(phi+alpha-beta) | 1
-        BSgate(pi/4, pi/2) | 1, 2
-        Rgate(2*theta+pi) | 1, 2
-        BSgate(pi/4, pi/2) | 1, 2
-        Rgate(beta-theta+pi) | 1
-        Rgate(beta-theta) | 2
+    if nmax != mmax:
+        raise ValueError("U must be a square matrix")
+
+    if U[n, m] == 0:
+        phi_i = np.pi
+        phi_e = 0
+    elif U[n - 1, m] == 0:
+        phi_i = 0
+        phi_e = 0
+    else:
+        r = U[n - 1, m] / U[n, m]
+        phi_i = 2 * np.arctan(np.abs(r))
+        phi_e = -np.angle(r)
+
+    return [n - 1, n, phi_i, phi_e, nmax]
 
-    The phase-shifts by ``alpha`` and ``beta`` are thus pushed consecutively through
-    all the T unitaries of the interferometer and these unitaries are converted
-    into pairs of symmetric beamsplitters with two phase shifts. The phase
-    shifts at the end of the interferometer are added to the ones from the
-    diagonal unitary at the end of the interferometer obtained from :func:`~.rectangular_phase_end`.
+
+def rectangular_MZ(V, tol=1e-11):
+    r"""Rectangular decomposition of a unitary matrix, with local
+    phase shifts applied between two interferometers.
+
+    Is similar to :func:`rectangular` except that it uses mach_zehner matrices to null elements of V
+    using the :func:`null_MZ` and :func:`null_MZi` instead of :func:`T` matrices and corresponding :func:`nullT`
+    and :func:`nullTi` functions.
```

Comment:
```suggestion
    Is similar to :func:`~.rectangular` except that it uses Mach Zehnder matrices to null elements of V
    using the :func:`~.null_MZ` and :func:`~.null_MZi` instead of :func:`~.T` matrices and corresponding :func:`~.nullT`
    and :func:`~.nullTi` functions.
```

Unless otherwise specified, Sphinx will assume the function/class name is an absolute path. If no function/class exists at that absolute path, sphinx will just silently fail to resolve the link.

We can use two syntactical tricks here to avoid this:

* Pattern matching (`.`): resolve to a class or function at any path matching a particular name.

  Note: in the case where there exists two objects with the same name in different places, parts of the path can be specified to avoid ambiguity. For example:

  ```rest
  :func:`.utils.catstate` and :func:`.ops.catstate`.
  ```

* When using `.` to pattern match, Sphinx will display the absolute path in the rendered link. For example, the code block above will render as

  `strawberryfields.utils.catstate` and `strawberryfields.ops.catstate`.

  To avoid this, we can use the prefix `~` to indicate to Sphinx to only use the last component of the target in the link text:

  ```rest
  :func:`~.utils.catstate` and :func:`~.ops.catstate`.
  ```